### DISCUSSION TITLE
How to Retrieve original file name iOS

### DISCUSSION THREAD

**Discussion by parthjdabhi** [react-native-image-picker/react-native-image-picker#2269] [General]
There is an issue in [react-native-image-picker](https://github.com/react-native-image-picker/react-native-image-picker)  v5.7.0, That we not getting an original filename instead we gets random UUID-based filename.

Just change this code to start getting the original filename in iOS.
Replace following code in _node_modules/react-native-image-picker/ios/**ImagePickerManager.mm**_

```
- (NSString *)getImageFileName:(NSString *)fileType
{
    NSString *fileName = [[NSUUID UUID] UUIDString];
    fileName = [fileName stringByAppendingString:@"."];
    return [fileName stringByAppendingString:fileType];
}
```
With following
```
- (NSString *)getImageFileNameFrom:(PHAsset * _Nullable)phAsset ForType:(NSString *)fileType
{
    if (phAsset && [phAsset valueForKey:@"filename"]) {
        NSString *name = [phAsset valueForKey:@"filename"];
        // Replace .HEIC extension with .jpg
        if ([name hasSuffix:@"HEIC"]) {
          name = [name stringByReplacingOccurrencesOfString:@"HEIC" withString:@"jpg"];
        }
        return name;
    }
    NSString *fileName = [[NSUUID UUID] UUIDString];
    fileName = [fileName stringByAppendingString:@"."];
    return [fileName stringByAppendingString:fileType];
}
```

and change `[self getImageFileName:fileType];` to `[self getImageFileNameFrom:phAsset ForType:fileType];`

That's it.

Run an iOS app again and you will start getting an original file name in fileName key.

![Screenshot 2024-01-31 at 7 14 41 PM](https://github.com/react-native-image-picker/react-native-image-picker/assets/10777063/31c521e3-13f1-418f-b600-3bee9ecfb554)

**Comment by eduardo-santos-tribia** [comment]
Hi @parthjdabhi, it didn't work for me, but this other solution works and get the real file name.

```objectivec
- (NSString *)getImageFileNameFrom:(PHAsset * _Nullable)phAsset ForType:(NSString *)fileType
{
    if (phAsset) {
        NSArray<PHAssetResource *> *resources = [PHAssetResource assetResourcesForAsset:phAsset];
        if (resources.count > 0) {
            NSString *name = resources.firstObject.originalFilename;
            // Replace .HEIC extension with .jpg
            if ([name hasSuffix:@"HEIC"]) {
                name = [name stringByReplacingOccurrencesOfString:@"HEIC" withString:@"jpg"];
            }
            return name;
        }
    }
    
    NSString *fileName = [[NSUUID UUID] UUIDString];
    fileName = [fileName stringByAppendingString:@"."];
    return [fileName stringByAppendingString:fileType];
}
```

Result:
```json
{
  "asset": {
    "fileName": "JPEG sample.jpeg",
    "fileSize": 9166,
    "height": 183,
    "id": "1FA1D528-72ED-450C-88F4-AADFB74B49C0/L0/001",
    "timestamp": "2023-04-27T15:00:36.073+0200",
    "type": "image/jpg",
    "uri": "file:///Users/myUser/Library/Developer/CoreSimulator/Devices/971B3453-A96D-425B-9828-12B27F096F2F/data/Containers/Data/Application/0A19972A-C12D-4C59-8081-1B50C5F3C921/tmp/JPEG%20sample.jpeg",
    "width": 275
  }
}
```

Now I'd like to do the same for Android...

**Reply by bormm** [reply]
Shouldn't this instead
`name = [name stringByReplacingOccurrencesOfString:@"HEIC" withString:@"jpg"];`

be like this:
`name = [name stringByReplacingOccurrencesOfString:@"HEIC" withString:fileType];`
?

**Comment by DeepakSharma04** [comment]
> Hi @parthjdabhi, it didn't work for me, but this other solution works and get the real file name.
> 
> ```objc
> - (NSString *)getImageFileNameFrom:(PHAsset * _Nullable)phAsset ForType:(NSString *)fileType
> {
>     if (phAsset) {
>         NSArray<PHAssetResource *> *resources = [PHAssetResource assetResourcesForAsset:phAsset];
>         if (resources.count > 0) {
>             NSString *name = resources.firstObject.originalFilename;
>             // Replace .HEIC extension with .jpg
>             if ([name hasSuffix:@"HEIC"]) {
>                 name = [name stringByReplacingOccurrencesOfString:@"HEIC" withString:@"jpg"];
>             }
>             return name;
>         }
>     }
>     
>     NSString *fileName = [[NSUUID UUID] UUIDString];
>     fileName = [fileName stringByAppendingString:@"."];
>     return [fileName stringByAppendingString:fileType];
> }
> ```
> 
> Result:
> 
> ```json
> {
>   "asset": {
>     "fileName": "JPEG sample.jpeg",
>     "fileSize": 9166,
>     "height": 183,
>     "id": "1FA1D528-72ED-450C-88F4-AADFB74B49C0/L0/001",
>     "timestamp": "2023-04-27T15:00:36.073+0200",
>     "type": "image/jpg",
>     "uri": "file:///Users/myUser/Library/Developer/CoreSimulator/Devices/971B3453-A96D-425B-9828-12B27F096F2F/data/Containers/Data/Application/0A19972A-C12D-4C59-8081-1B50C5F3C921/tmp/JPEG%20sample.jpeg",
>     "width": 275
>   }
> }
> ```
> 
> Now I'd like to do the same for Android...

##
Hi @parthjdabhi , can you provide the code changes for android to get the real file name .

**Reply by parthjdabhi** [reply]
@DeepakSharma04 just update **Utils.java** file at _node_modules/react-native-image-picker/android/src/main/java/com/imagepicker/Utils.java_ a path to this file to get original file name in android as well.
https://gist.github.com/parthjdabhi/01027b4a677a161f5593d28830c39b34

**Reply by eduardo-santos-tribia** [reply]
@parthjdabhi the `Utils.java` solution is not working for me. It's still creating the file with a strange name like "12300001.jpg" for example...

**Reply by VictorDev200** [reply]
Hello @parthjdabhi , i believe you uploaded the wrong file, because that file only has changes related to the dimensions of the image (comparing it to the original Utils on 7.1.2). Can please review it and upload the solution to android, that would help us a lot :D. Also its not too much to ask, can you make the same for the videos? to get the real URL of them. thanks :D

**Reply by DeepakSharma04** [reply]
use this - https://github.com/ivpusic/react-native-image-crop-picker

**Reply by VictorDev200** [reply]
@DeepakSharma04 Im still getting cache paths with that library:

"path": "file:///storage/emulated/0/Android/data/com.myApp/cache/temp/1712907365187.jpg"

**Reply by DeepakSharma04** [reply]
use patch packages --https://github.com/ivpusic/react-native-image-crop-picker/issues/1078#issuecomment-1609702415

**Reply by VictorDev200** [reply]
[{"filename": "1712909182355.jpg", "height": 2448, "mime": "image/jpeg", "modificationDate": "1712909182354", "path": "file:///storage/emulated/0/Android/data/com.myApp/cache/temp/1712909182355.jpg", "size": 633090, "width": 2448}]

:D (going to madness)

The image name is 20240412_010615.jpg

**Reply by VictorDev200** [reply]
````java
diff --git a/node_modules/react-native-image-crop-picker/android/src/main/java/com/reactnative/ivpusic/imagepicker/PickerModule.java b/node_modules/react-native-image-crop-picker/android/src/main/java/com/reactnative/ivpusic/imagepicker/PickerModule.java
index 5de0845..457de06 100644
--- a/node_modules/react-native-image-crop-picker/android/src/main/java/com/reactnative/ivpusic/imagepicker/PickerModule.java
+++ b/node_modules/react-native-image-crop-picker/android/src/main/java/com/reactnative/ivpusic/imagepicker/PickerModule.java
@@ -692,6 +692,7 @@ class PickerModule extends ReactContextBaseJavaModule implements ActivityEventLi
         image.putString("mime", options.outMimeType);
         image.putInt("size", (int) new File(compressedImagePath).length());
         image.putString("modificationDate", String.valueOf(modificationDate));
+        image.putString("filename", new File(path).getName());
         
         if (includeBase64) {
             image.putString("data", getBase64StringFromFile(compressedImagePath));

````

**Reply by DeepakSharma04** [reply]
**Android**
`[{"filename": "Rose.jpg", "height": 2736, "mime": "image/jpeg", "modificationDate": "1712913777000", "path": "file:///data/user/0/com.faveo.helpdesk.rn/cache/react-native-image-crop-picker/Rose.jpg", "size": 331466, "width": 3648}]`

**ios**
`[{"creationDate": "1712818241", "cropRect": null, "data": null, "duration": null, "exif": null, "filename": "IMG_0009.JPG", "height": 2736, "localIdentifier": "A2EA4B52-F9FA-47B6-B173-1275330B9799/L0/001", "mime": "image/jpeg", "modificationDate": "1712818241", "path": "/Users/deepaksharma/Library/Developer/CoreSimulator/Devices/2F0123F9-E230-43FB-A094-93E22F14B55A/data/Containers/Data/Application/E365F5B0-F992-47F8-9545-9E9B5308B9FF/tmp/react-native-image-crop-picker/1BC38EE5-B613-4C28-A47B-1FCF77D765C5.jpg", "size": 331658, "sourceURL": "file:///Users/deepaksharma/Library/Developer/CoreSimulator/Devices/2F0123F9-E230-43FB-A094-93E22F14B55A/data/Media/DCIM/100APPLE/IMG_0009.JPG", "width": 3648}]`

**Reply by VictorDev200** [reply]
@DeepakSharma04 maybe because i still have react-native-image-picker installed?

**Reply by DeepakSharma04** [reply]
Yes , remove it and install latest react-native-image-crop-picker with patch the android file.It will give correct name .

**Comment by bormm** [comment]
@parthjdabhi 

I am not sure if calling this "original filename" is correct. I have no iOS device currently and only worked with iOS some years ago, but isn't it true that the camera only stores heic files since a while? So the original file is always a heic file and its only transcoded to jpg by iOS if the preferredAssetRepresentationMode is set to not set to PHPickerConfigurationAssetRepresentationModeCurrent? 

This means this picker can't return the correct original filename at all, because nobody want the original name really.

That would also explain the needed replacement of the heic extension with jpg that looks strange in all the suggested fixes.

**Comment by JuliaSzymanskaFirmao** [comment]
Hi, I improved upper solutions, because they didn't work for me. Here is my patch file to get the initial filename for iOS:

[react-native-image-picker+5.7.0.patch](https://github.com/user-attachments/files/15836511/react-native-image-picker%2B5.7.0.patch)

**Reply by eduardo-santos-tribia** [reply]
@JuliaSzymanskaFirmao could you create a pull request with the fix? Then we can make it available to everyone.
